core/src/main/scala/slamdata/AutoBump.scala

```diff
@@ -0,0 +1,369 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package precog
+import java.io.File
+import java.nio.file.Files
+
+import org.http4s.Uri
+
+import cats.Monad
+import cats.data.EitherT
+import cats.effect.IO.contextShift
+import cats.effect.{ContextShift, IO, Sync}
+import cats.implicits._
+import github4s.GithubResponses.{GHException, GHResponse, GHResult}
+import github4s.domain._
+import precog.domain.{PullRequestDraft, PullRequestUpdate}
+import sbt.{Logger, url}
+import sbttrickle.metadata.OutdatedRepository
+import fs2.{Chunk, Stream}
+
+import scala.collection.immutable.Seq
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.sys.process._
+import scala.util.Try
+import scala.util.matching.Regex
+
+object AutoBump {
+  sealed trait Warnings {
+    def warn(log: Logger): IO[Unit]
+  }
+  case object UpdateError extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn("was unable to run `sbt update` following the trickle application")
+      log.warn("this may mean that the some artifacts are not yet propagated; skipping")
+    }
+  }
+  case object NoChangesError extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn("git-commit exited with error")
+      log.warn("this usually means the target repository was *already* at the latest version but hasn't published yet")
+      log.warn("you should check for a stuck trickle PR on that repository")
+    }
+  }
+  case object PushError extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn("git-push exited with error")
+      log.warn("this usually means some other repository updated the pull request before this one")
+    }
+  }
+  case class NotOldest(oldest: PullRequestDraft, draft: PullRequestDraft) extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn(s"pull request ${draft.number} is newer than existing pull request ${oldest.number}")
+      log.warn("this usually means two or more repositories finished build at the same time,")
+      log.warn("and some other repository beat this one to pull request creation.")
+    }
+  }
+
+  sealed abstract class ChangeLabel(val label: String)
+  object ChangeLabel {
+    val values: List[ChangeLabel] = List(RevisionChange, FeatureChange, BreakingChange)
+    val order: Map[ChangeLabel, Int] = values.zipWithIndex.toMap
+    implicit val ordering: Ordering[ChangeLabel] = Ordering.by(order)
```

Comment:
Minor aside: we have Cats in scope here. Rather than using `Ordering`, it's probably better to just use `Order`. I'm not super-fussed about it either way.

Reply:
My mind was about to go into shutdown, and I'm not familiar enough with Cats yet. Hell, even dragging out the details of `Ordering` was a major ordeal at that point.